src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -0,0 +1,227 @@
+
+        private VSClientCapabilities? _clientCapabilities;
+
+        public InProcLanguageServer(Stream inputStream, Stream outputStream, LanguageServerProtocol protocol, Workspace workspace, IDiagnosticService diagnosticService, IThreadingContext threadingContext)
+        {
+            this._protocol = protocol;
+            this._workspace = workspace;
+            this._threadingContext = threadingContext;
+
+            this._jsonRpc = new JsonRpc(outputStream, inputStream, this);
+            this._jsonRpc.StartListening();
+
+            this._diagnosticService = diagnosticService;
+            this._diagnosticService.DiagnosticsUpdated += DiagnosticService_DiagnosticsUpdated;
+        }
+
+        /// <summary>
+        /// Handle the LSP initialize request by storing the client capabilities
+        /// and responding with the server capabilities.
+        /// The specification assures that the initialize request is sent only once.
+        /// </summary>
+        [JsonRpcMethod(Methods.InitializeName)]
+        public InitializeResult Initialize(JToken input)
+        {
+            // InitializeParams only references ClientCapabilities, but the VS LSP client
+            // sends additional VS specific capabilities, so directly deserialize them into the VSClientCapabilities
+            // to avoid losing them.
+            this._clientCapabilities = input["capabilities"].ToObject<VSClientCapabilities>();
+
+            return new InitializeResult
+            {
+                Capabilities = new VSServerCapabilities
+                {
+                    DocumentHighlightProvider = true,
+                    DocumentSymbolProvider = true,
+                    DocumentFormattingProvider = true,
+                    DocumentRangeFormattingProvider = true,
+                    DocumentOnTypeFormattingProvider = new DocumentOnTypeFormattingOptions { FirstTriggerCharacter = "}", MoreTriggerCharacter = new[] { ";", "\n" } },
+                    DefinitionProvider = true,
+                    CompletionProvider = new CompletionOptions { ResolveProvider = true, TriggerCharacters = new[] { "." } },
+                    ImplementationProvider = true,
+                    ReferencesProvider = true,
+                    SignatureHelpProvider = new SignatureHelpOptions { TriggerCharacters = new[] { "(", "," } },
+                    WorkspaceSymbolProvider = true,
+                }
+            };
+        }
+
+        [JsonRpcMethod(Methods.InitializedName)]
+        public Task Initialized() => Task.CompletedTask;
+
+        [JsonRpcMethod(Methods.ShutdownName)]
+        public object? Shutdown(CancellationToken _) => null;
+
+        [JsonRpcMethod(Methods.ExitName)]
+        public void Exit()
+        {
+        }
+
+        [JsonRpcMethod(Methods.TextDocumentDefinitionName)]
+        public async Task<object> GetTextDocumentDefinitionAsync(JToken input, CancellationToken cancellationToken)
```

Comment:
Does this one also need to run on the UI thread? It eventually does some stuff similar to FAR where it makes an rqname and broadcasts it to the world (http://sourceroslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Workspace/VisualStudioSymbolNavigationService.cs,a5ffa89c5caac2ee)

Reply:
:( yup it looks like it.